src/main/java/com/mongodb/spark/sql/connector/read/partitioner/PartitionerHelper.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.mongodb.spark.sql.connector.read.partitioner;
+
+import static com.mongodb.spark.sql.connector.read.partitioner.Partitioner.LOGGER;
+import static java.util.Collections.singletonList;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.bson.BsonDocument;
+import org.bson.BsonType;
+import org.bson.BsonValue;
+
+import com.mongodb.MongoCommandException;
+import com.mongodb.client.MongoClient;
+
+import com.mongodb.spark.sql.connector.config.ReadConfig;
+import com.mongodb.spark.sql.connector.exceptions.MongoSparkException;
+import com.mongodb.spark.sql.connector.read.MongoInputPartition;
+
+/** The Partitioner helper class */
+public final class PartitionerHelper {
+
+  private static final List<BsonDocument> COLL_STATS_AGGREGATION_PIPELINE =
+      singletonList(BsonDocument.parse("{'$collStats': {'storageStats': { } } }"));
+  public static final String DEFAULT_PARTITIONER = SamplePartitioner.class.getCanonicalName();
+  public static final Partitioner SINGLE_PARTITIONER = new SinglePartitionPartitioner();
```

Comment:
All partitioners will only be called by a single Spark driver.

Reply:
I don't think I understand why all partitioners are now required to be thread-safe. As far as I can see, partitioners are instantiated only in `PartitionerHelper.generatePartitions` and are thrown away when the method returns. This means that no partitioner instantiated there is ever be used concurrently, i.e., they don't need to be thread-safe.

`SinglePartitionPartitioner` is the only partitioner which instance may in principle be used concurrently, because we create a single instance referred to by a static field. That is why I commented that we seem to rely on the thread-safety of this class.

TL;DR I was talking about `SinglePartitionPartitioner `, and see no reason to require all partitioners to be thread-safe: it is unhelpful to us, and further harmful for those who may want to implement a custom partitioner.

Reply:
Ack - updated.